library_management_system_downloader/create_api_request.py

```diff
@@ -0,0 +1,51 @@
+import backoff
+from ratelimit import rate_limited
+import requests
+import logging
+
+
+class ErrorWithAPI(Exception):
+    pass
+
+
+# From the backoff documentation (https://pypi.python.org/pypi/backoff),
+# set maximum number of tries on a failed download:
+@backoff.on_exception(
+        backoff.expo,
+        requests.exceptions.RequestException,
+        max_tries=8)
+# From the ratelimit documentation (https://pypi.python.org/pypi/ratelimit),
+# set the rate limit (in seconds) for API calls:
+@rate_limited(0.2)
+def create_api_request(
+        item_doi,
+        api_base_url,
+        static_api_request_parameters_dictionary=None,
+        custom_user_agent_string=None):
+    """Given an item DOI (lacking the 'doi:/' prefix), query an OpenURL
+    resolver) and return XML from it."""
+
+    if static_api_request_parameters_dictionary is None:
+        static_api_request_parameters_dictionary = {}
+
+    custom_api_query_header = {}
+    if custom_user_agent_string:
+        custom_api_query_header['user-agent'] = custom_user_agent_string
+
+    # Update the static api parameters to include the (dynamic) DOI:
+    api_request_parameters = static_api_request_parameters_dictionary.copy()
```

Comment:
Part 3: replace `static_api_request_parameters_dictionary` with `api_request_parameters`.

Better to avoid using additional variable names in this case, since we will never want access to the original (pre-copy) dictionary.

Reply:
So that line would read `api_request_parameters = api_request_parameters.copy()`? Or `api_request_parameters = api_request_parameters`?

Maybe I've been away from Python for too long, but I'm still not understanding this approach, sorry. (I'm not arguing against it; I'm just trying to understand it). My understanding is that setting a default parameter as `{}` is often undesirable since it can behave in ways that users don't expect. So I'm trying to understand why it *is* desirable here (since you suggested it).

Just making sure: The configuration variable is called `static_parameters_for_api` (not `static_api_request_parameters_dictionary`, which is what's in the code line referenced above). Do you mean for this line to read `api_request_parameters = static_parameters_for_api.copy()`?

Sorry if this seems (or is) dense; I appreciate your feedback on this approach.

Reply:
This line would become

```python
api_request_parameters = api_request_parameters.copy()
```

My suggestion is to use one variable name to refer to the dictionary prior to copying and the dictionary post copying. They will be different objects in memory.

However, performing the copy ensures that you don't edit a user-provided dict or the default parameter empty dict.

According to my suggestion `static_api_request_parameters_dictionary` will no longer appear in the code.

Reply:
I appreciate your patience and taking the time to explain; thank you. Implemented in 6d6c746a6.